core-bundle/src/Resources/contao/modules/ModuleLogin.php

```diff
@@ -65,14 +66,28 @@ public function generate()
 		{
 			$this->targetPath = base64_decode($request->request->get('_target_path'));
 		}
-		elseif ($this->redirectBack && $request && $request->query->has('redirect'))
+		elseif ($request && $this->redirectBack)
 		{
-			$uriSigner = System::getContainer()->get('uri_signer');
+			if ($request->query->has('redirect'))
+			{
+				$uriSigner = System::getContainer()->get('uri_signer');
 
-			// We cannot use $request->getUri() here as we want to work with the original URI (no query string reordering)
-			if ($uriSigner->check($request->getSchemeAndHttpHost() . $request->getBaseUrl() . $request->getPathInfo() . (null !== ($qs = $request->server->get('QUERY_STRING')) ? '?' . $qs : '')))
+				// We cannot use $request->getUri() here as we want to work with the original URI (no query string reordering)
+				if ($uriSigner->check($request->getSchemeAndHttpHost() . $request->getBaseUrl() . $request->getPathInfo() . (null !== ($qs = $request->server->get('QUERY_STRING')) ? '?' . $qs : '')))
+				{
+					$this->targetPath = $request->query->get('redirect');
+				}
+			}
+			elseif ($referer = $request->headers->get('referer'))
 			{
-				$this->targetPath = $request->query->get('redirect');
+				$refererUri = new Uri($referer);
+				$requestUri = new Uri($request->getUri());
+
+				// Use the HTTP referer as a fallback, but only if scheme and host matches with the current request (see #5860)
+				if ($refererUri->getScheme() === $requestUri->getScheme() && $refererUri->getHost() === $requestUri->getHost() && $refererUri->getPort() === $requestUri->getPort())
+				{
+					$this->targetPath = $referer;
```

Comment:
```suggestion
					$this->targetPath = (string) $refererUri;
```
The parsed version needs to be used in order to guarantee that all special characters are encoded AFAIK.

Reply:
For example:
```php
(string) (new Uri('https://example.com/<script>')) === 'https://example.com/%3Cscript%3E'
```

Reply:
I'll do a follow up PR 👍